src/net.md

```diff
@@ -1019,6 +1020,151 @@ fn run() -> Result<()> {
 # quick_main!(run);
 ```
 
+[ex-progress-with-range]: #ex-progress-with-range
+<a name="ex-progress-with-range"/>
+## Make a partial download with HTTP range headers
+
+[![reqwest-badge]][reqwest] [![cat-net-badge]][cat-net]
+
+Uses [`reqwest::Client::head`] to get the content-length and validate if the server sets the header
+[`reqwest::header::ContentRange`], required to confirm the support of partial downloads.
+
+If supported downloads the content using [`reqwest::get`], setting the [`reqwest::header::Range`]
+to do partial downloads in chunks of 100 bytes, between these writes basic progress messages.
+
+Range header, defined in [RFC7233][HTTP Range RFC7233].
+
+```rust,no_run
+# #[macro_use]
+# extern crate error_chain;
+extern crate reqwest;
+
+use std::io::{Read, Write};
+use std::fs::File;
+
+use reqwest::header::{ContentRange, ContentRangeSpec, Range};
+
+# error_chain! {
+#     foreign_links {
+#         Io(std::io::Error);
+#         Reqwest(reqwest::Error);
+#     }
+# }
+
+const BUFFER_SIZE: u64 = 100;
+
+struct PartialRange {
```

Comment:
I propose to merge the `PartialRange` and `PartialRangeIter` into one which will be much more concise.

All we will need will be:

- the struct 
```rust
struct PartialRangeIter {
    start: u64,  // content range start
    end: u64,   // content range end
    buffer_size: usize, // external buffer size 
}
```
- The constructor `new` function below
- iter impl which will be much simplified

We should not need anything more complicated than
```rust
impl Iterator for PartialRangeIter {
    type Item = Range;

    fn next(&mut self) -> Option<Self::Item> {
        if self.start >= self.end {
            None
        } else {
            let old_start = self.start;
            self.start += cmp::min(self.buffer_size as u64, self.end - self.start +1);
            Some(Range::bytes(old_start, self.start -1))
        }
    }
}
```

Reply:
I have tried that approach during this end up with the following error, and now is back again with this code you're posting here:

```text
error[E0599]: no method named `iter` found for type `PartialRangeIter` in the current scope
  --> src/main.rs:77:25
   |
77 |     for range in ranges.iter() {
   |                         ^^^^
```

here is the complete `struct` + `impl`:

```rust
#[derive(Debug)]
struct PartialRangeIter {
    start: u64, // content range start
    end: u64, // content range end
    buffer_size: usize, // external buffer size
}

impl PartialRangeIter {
    pub fn new(content_range: &ContentRangeSpec, buffer_size: usize) -> Result<PartialRangeIter> {
        if buffer_size == 0 {
            Err("invalid buffer_size, give a value greater than zero.")?;
        }

        let (start, end) = match *content_range {
            ContentRangeSpec::Bytes { range: Some(range), .. } => (range.0, range.1),
            _ => Err("invalid range specification")?,
        };

        Ok(PartialRangeIter {
            start,
            end,
            buffer_size,
        })
    }
}

impl Iterator for PartialRangeIter {
    type Item = Range;

    fn next(&mut self) -> Option<Self::Item> {
        if self.start >= self.end {
            None
        } else {
            let old_start = self.start;
            let inc = std::cmp::min(self.buffer_size as u64, self.end - self.start + 1);
            self.start += std::cmp::min(self.buffer_size as u64, self.end - self.start + 1);
            Some(Range::bytes(old_start, self.start - 1))
        }
    }
}
```

can you explain?

Reply:
> for range in ranges.iter() {

There is nothing to provide the `iter` method nor is it required.
Som types provide `iter`, `into_iter` or both to return a separate helper iterator type.
In this situation we already have an iterator type instance as it can be constructed with `PartialRangeIter::new` (not being tied to any other type).

So following would simply work 

```rust
for range in ranges {
```

I hope that helpes :+1: